tools/system_libs.py

```diff
@@ -1662,7 +1662,7 @@ def calculate(input_files, forced):
   force = os.environ.get('EMCC_FORCE_STDLIBS')
   if force == '1':
     force = ','.join(name for name, lib in system_libs_map.items() if not lib.never_force)
-  force_include = set((force.split(',') if force else []) + forced)
+  force_include = sorted(set((force.split(',') if force else []) + forced))
```

Comment:
I think the sorting is only needed in the `=1` case (i.e. when all libraries are requested).  How about this instead:

```
if force == '1':
  all_names = sorted(name for name, lib in system_libs_map.items() if not lib.never_force)
  force = ','.join(all_names)
```

I do wonder if the sort order matters here if we should thing about an correct ordering .. but on the other hand I can't believe that `EMCC_FORCE_STDLIBS=1` is very commonly used.

Reply:
Actually maybe we should just both to join() just so can we spit on the next line. and also maybe we should avoid all this logic when EMCC_FORCE_STDLIBS is not part of the environment (the most likely case).

How about we land this change, but then as a followup we refactor to be more like this:

```
user_forced = []
if 'EMCC_FORCE_STDLIBS' in os.environ:
  force = os.environ.get('EMCC_FORCE_STDLIBS')
  if force == '1':
    user_forced = sorted(name for name, lib in system_libs_map.items() if not lib.never_force)
  else:
    user_forced = force.split(',')
force_include = user_forced + forced
```

Reply:
The first suggestion didn't work, probably because I'm not using `EMCC_FORCE_STDLIBS`. I did this quick test:
```diff
--- a/tools/system_libs.py
+++ b/tools/system_libs.py
@@ -1662,8 +1662,9 @@ def calculate(input_files, forced):
   force = os.environ.get('EMCC_FORCE_STDLIBS')
   if force == '1':
     force = ','.join(name for name, lib in system_libs_map.items() if not lib.never_force)
-  force_include = sorted(set((force.split(',') if force else []) + forced))
+  force_include = set((force.split(',') if force else []) + forced)
   if force_include:
+    logger.debug(f'{forced}')
     logger.debug(f'forcing stdlibs: {force_include}')
 
   def add_library(libname):
```

And noticed:
```
system_libs:DEBUG: ['libwasmfs', 'libembind']
system_libs:DEBUG: forcing stdlibs: {'libembind', 'libwasmfs'}
```

So the `set()` causes the order to be lost. Perhaps we should use `OrderedDict` instead? That will remember insertion order.
```diff
--- a/tools/system_libs.py
+++ b/tools/system_libs.py
@@ -11,6 +11,7 @@ import os
 import shutil
 from enum import IntEnum, auto
 from glob import iglob
+from collections import OrderedDict
 
 from . import shared, building, utils
 from . import deps_info, tempfiles
@@ -1662,7 +1663,7 @@ def calculate(input_files, forced):
   force = os.environ.get('EMCC_FORCE_STDLIBS')
   if force == '1':
     force = ','.join(name for name, lib in system_libs_map.items() if not lib.never_force)
-  force_include = sorted(set((force.split(',') if force else []) + forced))
+  force_include = list(OrderedDict.fromkeys((force.split(',') if force else []) + forced))
   if force_include:
     logger.debug(f'forcing stdlibs: {force_include}')
 ```

Of course, if we don't need to remove duplicates we can just remove this `set()` (like your second suggestion).

Reply:
Hmm, perhaps we should prefer insertion order here, since ports are linked in dependency order.
https://github.com/emscripten-core/emscripten/blob/a8e7d3b8f27b05eb39b73151e1ff2914241acae8/emcc.py#L2613-L2617

Probably a `sorted()` would break that.

Reply:
Oh I see.  Yes I think just making it a list should be fine. It also fine if it has duplicates I think.. so just removing the set() call should be enough:

```
force_include = (force.split(',') if force else []) + forced
```

Reply:
Also a nice simplifcation would be to just do: 

```
force = os.environ.get('EMCC_FORCE_STDLIBS'. '')
...
force_include = force.split(',') + forced
```

Reply:
OK, PR has been updated to remove the `set()` in favor of a list. I also simplified the `EMCC_FORCE_STDLIBS` logic a bit.